### PR TITLE
Dynamic API

### PR DESCRIPTION
**Pull request by Vijar**
@redonkulus @jedireza

### CODE REVIEW

File: secrets.example.js [hunk 1]
```diff
@@ -1,6 +0,0 @@
-module.exports = {
```
Comment (redonkulus) [inline secrets.example.js]: why was this deleted?

File: package.json [hunk 2]
```diff
@@ -48,8 +48,10 @@
     "marked": "^0.3.2",
     "node-jsx": "^0.12.0",
     "react": "^0.12.0",
+    "request": "^2.51.0",
```
Comment (redonkulus) [inline package.json]: is this needed?

**Comment by redonkulus** [comment]
Start of this looks good but we need to abstract only the API info from the readme's and not just display the whole thing.

File: secrets.example.js (hunk 1 above)
Comment (Vijar) [inline secrets.example.js]: I moved it to secrets.js, but looks like secrets.js was ignored by .gitignore

Whats the best approach here?

File: package.json (hunk 2 above)
Comment (Vijar) [inline package.json]: Nope. Initially used request, and then switched to superagent.

File: secrets.example.js (hunk 1 above)
Comment (jedireza) [inline secrets.example.js]: Yes we ignore it so we don't commit our secrets to the repo. This example file should stay though.

**Comment by Vijar** [comment]
Yeah, I was thinking about that.

The best method to extract part of the markdown file might be to surround the API section with ["markdown comments"](http://stackoverflow.com/a/20885980) 

```
[//]: # (API_START)
# API Stuff
[//]: # (API_STOP)
```

[//]: # (API_START)
# API Stuff
[//]: # (API_STOP)

**Comment by redonkulus** [comment]
The comments approach sounds good.

**Comment by Vijar** [comment]
@redonkulus @mridgway

File: services/api.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2014, Yahoo! Inc.
```
Comment (redonkulus) [inline services/api.js]: 2015

File: services/api.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    var key = 'api/' + repo;
+    var url = 'https://api.github.com/repos/yahoo/';
+    url += repo;
+    url += '/readme';
+    url += '?' + qs.stringify({
+        client_id: secret.github.clientId,
+        client_secret: secret.github.clientSecret
+    });
+
+    var extractAPI = function (readme) {
```
Comment (redonkulus) [inline services/api.js]: Minor: could move this out of the function so v8 can optimize it and optimize fetchReadme.

File: services/api.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    setTimeout(refreshCacheFromGithub, 1 * 60 * 60 * 1000); // refresh cache every hour
+})();
+
+module.exports = {
+    name: 'api',
+    read: function (req, resource, params, config, callback) {
+        // Throw an error if repo is not in whitelist
+        if (whitelistedRepos.indexOf(params.repo) === -1) {
+            return callback(new Error('Repo not whitelisted: ' + params.repo));
```
Comment (redonkulus) [inline services/api.js]: If this is user facing, they don't know its a repo. Could just say "API not found" or something.

File: components/Application.jsx [hunk 6]
```diff
@@ -49,7 +49,12 @@ var Application = React.createClass({
             hideLogo = true;
         }
         else if ('docs' === this.state.currentPageName) {
-            page = <Docs context={this.props.context} content={this.state.currentDoc.content} />;
+            var config = require('./../configs/docs')
+            page = <Docs menu={config} context={this.props.context} content={this.state.currentDoc.content} />;
+        }
+        else if ('apis' === this.state.currentPageName) {
+            var config = require('./../configs/apis')
```
Comment (jedireza) [inline components/Application.jsx]: FYI: as is we're missing semicolons.

But let's move `require` statements to the top. We just need to create two variables for the configs and use the right one accordingly.

```js
var configDocs = require('./../configs/docs');
var configApis = require('./../configs/apis');

...

if ('home' === this.state.currentPageName) {
    page = <Home context={this.props.context} content={this.state.currentDoc.content} />;
    hideLogo = true;
}
else if ('docs' === this.state.currentPageName) {
    page = <Docs menu={configDocs} context={this.props.context} content={this.state.currentDoc.content} />;
}
else if ('apis' === this.state.currentPageName) {
    page = <Docs menu={configApis} context={this.props.context} content={this.state.currentDoc.content} />;
}
```

File: configs/docs.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            {
+                label: 'Controller Views',
+                routeName: 'docs',
+                navParams: {
+                    type: 'guides',
+                    key: 'controller-views'
+                }
+            },
+            {
+                label: 'Services / Fetching Data',
```
Comment (jedireza) [inline configs/docs.js]: Let's rename this to "Data Services". (currently the label is long and wraps when not at the largest break point)

File: services/api.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2014, Yahoo! Inc.
+ * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
+ */
+'use strict';
+
+var debug = require('debug')('APIService');
+var marked = require('marked');
+var request = require('superagent');
+var secret = require('./../secrets');
```
Comment (jedireza) [inline services/api.js]: `secrets`

**Comment by jedireza** [comment]
Very nice. Left a couple comments.

File: components/Application.jsx (hunk 6 above)
Comment (Vijar) [inline components/Application.jsx]: Usually, I'd agree with you for libs and such. But in this case, I like having the require in the if block, because it makes it easy to see where that config is used.

Since this is used exactly once, I'd even favor [inlining the require](https://github.com/yahoo/flux-examples/blob/79b9620e2a063b953899457e2e0fe0c8c8a0bfef/chat/app.js#L18) but, mixing jsx with require looked goofy.

I'm all for renaming `config` to be more specific like you did though.

File: configs/docs.js (hunk 7 above)
Comment (Vijar) [inline configs/docs.js]: I like that.

**Comment by jedireza** [comment]
Looks good. Should we squash all of these?

**Comment by Vijar** [comment]
Yesh, will be swashing before merge

**Comment by redonkulus** [comment]
:shipit:
